File: Assignments/DSA/Assignment_3/5.py
def constructPath(L, grid, tail):

    (i, j) = tail
    path = [tail]

    while L[i][j]:
        if i - 1 >= 0 and L[i][j] - L[i - 1][j] == 1 and \
                abs(grid[i - 1][j] - grid[i][j]) == 1:
            path.append((i - 1, j))
            i = i - 1
        elif j - 1 >= 0 and L[i][j] - L[i][j - 1] == 1 and \
                abs(grid[i][j - 1] - grid[i][j]) == 1:
            path.append((i, j - 1))
            j = j - 1

    return path


def findMaxLengthSnakeSequence(grid):

    if not grid or not len(grid):
        return
    L = [[0 for x in range(len(grid))] for y in range(len(grid))]
    max_so_far = 0

    tail = None

    for i in range(len(grid)):
        for j in range(len(grid)):

            if i - 1 >= 0 and abs(grid[i - 1][j] - grid[i][j]) == 1:
                L[i][j] = L[i - 1][j] + 1
                if max_so_far < L[i][j]:
                    max_so_far = L[i][j]
                    tail = (i, j)

            if j - 1 >= 0 and abs(grid[i][j - 1] - grid[i][j]) == 1:

                L[i][j] = max(L[i][j], L[i][j - 1] + 1)
                if max_so_far < L[i][j]:
                    max_so_far = L[i][j]
                    tail = (i, j)
    return constructPath(L, grid, tail)


def printSnakeSequence(grid, path):
    if not grid or not len(grid):
        return
    print('The maximum length snake sequence ', end='')
    print(' — '.join([str(grid[x][y]) for x, y in path][::-1]))
    print('The length is', len(path) - 1)


if __name__ == '__main__':

    grid = [
        [7, 5, 2, 3, 1],
        [3, 4, 1, 4, 4],
        [1, 5, 6, 7, 8],
        [3, 4, 5, 8, 9],
        [3, 2, 2, 7, 6]
    ]

    path = findMaxLengthSnakeSequence(grid)
    printSnakeSequence(grid, path)

File: Assignments/DSA/Assignment_3/5.cpp
#include <iostream>
#include <list>

using namespace std;
#define M 4
#define N 4

struct Point
{
    int x, y;
};

list<Point> findPath(int grid[M][N], int mat[M][N],
                     int i, int j)
{
    list<Point> path;

    Point pt = {i, j};
    path.push_front(pt);

    while (grid[i][j] != 0)
    {
        if (i > 0 &&
            grid[i][j] - 1 == grid[i - 1][j])
        {
            pt = {i - 1, j};
            path.push_front(pt);
            i--;
        }
        else if (j > 0 &&
                 grid[i][j] - 1 == grid[i][j - 1])
        {
            pt = {i, j - 1};
            path.push_front(pt);
            j--;
        }
    }

    return path;
}

void findSnakeSequence(int mat[M][N])
{

    int lookup[M][N];

    memset(lookup, 0, sizeof lookup);

    int max_len = 0;

    int max_row = 0;
    int max_col = 0;

    for (int i = 0; i < M; i++)
    {
        for (int j = 0; j < N; j++)
        {

            if (i || j)
            {

                if (i > 0 &&
                    abs(mat[i - 1][j] - mat[i][j]) == 1)
                {
                    lookup[i][j] = max(lookup[i][j],
                                       lookup[i - 1][j] + 1);

                    if (max_len < lookup[i][j])
                    {
                        max_len = lookup[i][j];
                        max_row = i, max_col = j;
                    }
                }

                if (j > 0 &&
                    abs(mat[i][j - 1] - mat[i][j]) == 1)
                {
                    lookup[i][j] = max(lookup[i][j],
                                       lookup[i][j - 1] + 1);
                    if (max_len < lookup[i][j])
                    {
                        max_len = lookup[i][j];
                        max_row = i, max_col = j;
                    }
                }
            }
        }
    }

    cout << "Maximum length of Snake sequence is: "
         << max_len << endl;

    list<Point> path = findPath(lookup, mat, max_row,
                                max_col);

    cout << "Snake sequence is:";
    for (auto it = path.begin(); it != path.end(); it++)
        cout << endl
             << mat[it->x][it->y] << " ("
             << it->x << ", " << it->y << ")";
}

int main()
{
    int mat[M][N] =
        {
            {9, 6, 5, 2},
            {8, 7, 6, 5},
            {7, 3, 1, 6},
            {1, 1, 1, 7},
        };

    findSnakeSequence(mat);

    return 0;
}